src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs

```diff
@@ -847,5 +854,65 @@ protected virtual CultureInfo Culture
                 return Arguments.PreferredUILang ?? CultureInfo.CurrentUICulture;
             }
         }
+
+        private static void EmitDeterminismKey(CommandLineArguments args, string[] rawArgs, string baseDirectory, CommandLineParser parser)
+        {
+            var key = CreateDeterminismKey(args, rawArgs, baseDirectory, parser);
+            var filePath = Path.Combine(args.OutputDirectory, args.OutputFileName + ".key");
+            using (var stream = PortableShim.File.Create(filePath))
+            {
+                var bytes = Encoding.UTF8.GetBytes(key);
+                stream.Write(bytes, 0, bytes.Length);
+            }
+        }
+
+        /// <summary>
+        /// The string returned from this function represents the inputs to the compiler which impact determinism.  It is 
+        /// meant to be inline with the specification here:
+        /// 
+        ///     - https://github.com/dotnet/roslyn/blob/master/docs/compilers/Deterministic%20Inputs.md
+        /// 
+        /// </summary>
+        private static string CreateDeterminismKey(CommandLineArguments args, string[] rawArgs, string baseDirectory, CommandLineParser parser)
+        {
+            List<Diagnostic> diagnostics = new List<Diagnostic>();
+            List<string> flattenedArgs = new List<string>();
+            parser.FlattenArgs(rawArgs, diagnostics, flattenedArgs, null, baseDirectory);
+
+            var builder = new StringBuilder();
+            var name = !string.IsNullOrEmpty(args.OutputFileName)
+                ? Path.GetFileNameWithoutExtension(Path.GetFileName(args.OutputFileName))
+                : $"no-output-name-{Guid.NewGuid().ToString()}";
+
+            builder.AppendLine($"{name}");
+            builder.AppendLine($"Command Line:");
+            foreach (var current in flattenedArgs)
+            {
+                builder.AppendLine($"\t{current}");
+            }
+
+            builder.AppendLine("Source Files:");
+            var hash = new MD5CryptoServiceProvider();
+            foreach (var sourceFile in args.SourceFiles)
+            {
+                var sourceFileName = Path.GetFileName(sourceFile.Path);
+
+                string hashValue;
+                try
+                {
+                    var bytes = PortableShim.File.ReadAllBytes(sourceFile.Path);
+                    var hashBytes = hash.ComputeHash(bytes);
+                    var data = BitConverter.ToString(hashBytes);
+                    hashValue = data.Replace("-", "");
+                }
+                catch (Exception ex)
+                {
+                    hashValue = $"Could not compute {ex.Message}";
+                }
+                builder.AppendLine($"\t{sourceFileName} - {hashValue}");
+            }
+
+            return builder.ToString();
```

Comment:
What about referenced assemblies? If we don't record them, customers will complain we're not deterministic even though the difference is due to different referenced assemblies (e.g. on different build machines).